engine/resource/src/resource_mounts.cpp

```diff
@@ -0,0 +1,539 @@
+#include <dlib/log.h>
+#include <dlib/mutex.h>
+#include <dlib/sys.h>
+#include <algorithm> // std::sort
+
+namespace dmResourceMounts
+{
+
+const char* MOUNTS_FILENAME = "liveupdate.mounts";
+
+const int MAX_NAME_LENGTH = 64;
+
+struct ArchiveMount
+{
+    const char*                     m_Name;
+    dmResourceProvider::HArchive    m_Archive;
+    int                             m_Priority;
+    bool                            m_Persist;
+};
+
+struct ResourceMountsContext
+{
+    // The currently mounted archives, in sorted order
+    dmArray<ArchiveMount>           m_Mounts;
+    dmResourceProvider::HArchive    m_ResourceBaseArchive;
+    dmMutex::HMutex                 m_Mutex;
+};
+
+
+static dmResource::Result ProviderResultToResult(dmResourceProvider::Result result)
+{
+    switch(result)
+    {
+    case dmResourceProvider::RESULT_OK:         return dmResource::RESULT_OK;
+    case dmResourceProvider::RESULT_IO_ERROR:   return dmResource::RESULT_IO_ERROR;
+    case dmResourceProvider::RESULT_NOT_FOUND:  return dmResource::RESULT_RESOURCE_NOT_FOUND;
+    default:                                    return dmResource::RESULT_UNKNOWN_ERROR;
+    }
+}
+
+static dmResource::Result DestroyMounts(HContext ctx);
+
+HContext Create(dmResourceProvider::HArchive base_archive)
+{
+    ResourceMountsContext* ctx = new ResourceMountsContext;
+    ctx->m_Mounts.SetCapacity(2);
+    ctx->m_Mutex = dmMutex::New();
+    ctx->m_ResourceBaseArchive = base_archive;
+    return ctx;
+}
+
+void Destroy(HContext ctx)
+{
+    {
+        DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+        DestroyMounts(ctx);
+    }
+    dmMutex::Delete(ctx->m_Mutex);
+    delete ctx;
+}
+
+struct MountSortPred
+{
+    bool operator ()(const ArchiveMount& a, const ArchiveMount& b) const
+    {
+        return a.m_Priority > b.m_Priority; // largest priority first
+    }
+};
+
+static void SortMounts(dmArray<ArchiveMount>& mounts)
+{
+    std::sort(mounts.Begin(), mounts.End(), MountSortPred());
+}
+
+static void AddMountInternal(HContext ctx, const ArchiveMount& mount)
+{
+    DM_MUTEX_SCOPED_LOCK(ctx->m_Mutex);
+
+    if (ctx->m_Mounts.Full())
+        ctx->m_Mounts.OffsetCapacity(2);
+
+    ctx->m_Mounts.Push(mount);
+    SortMounts(ctx->m_Mounts);
+
+    DM_RESOURCE_DBG_LOG(1, "Added archive %p with prio %d\n", mount.m_Archive, mount.m_Priority);
+}
+
+static void DebugPrintMount(int level, const ArchiveMount& mount)
+{
+    dmURI::Parts uri;
+    dmResourceProvider::GetUri(mount.m_Archive, &uri);
+    DM_RESOURCE_DBG_LOG(level, "Mount:  prio: %3d  %s  p: %p  uri: %s:%s/%s\n", mount.m_Priority, mount.m_Name, mount.m_Archive, uri.m_Scheme, uri.m_Location, uri.m_Path);
+}
+
+static void DebugPrintMounts(HContext ctx)
+{
+    DM_RESOURCE_DBG_LOG(2, "Mounts\n");
+    uint32_t size = ctx->m_Mounts.Size();
+    for (uint32_t i = 0; i < size; ++i)
+    {
+        DebugPrintMount(2, ctx->m_Mounts[i]);
+    }
+}
+
+dmMutex::HMutex GetMutex(HContext ctx)
+{
+    return ctx->m_Mutex;
+}
```

Comment:
One thing to look out for, is to make sure we protect the mounts/operations with the mutex.
For most operations, it is handled internally, but for GetDependencies(), we lock it from the call site.